ocaml/libs/stunnel/stunnel.ml

```diff
@@ -345,39 +344,51 @@ let with_attempt_one_connect ?unique_id ?(use_fork_exec_helper = true)
                    )
                  (path :: args)
               )
-          ) ;
+        in
         Unixfd.safe_close config_out ;
-        Unixfd.safe_close data_in ;
-        let config = config_file verify_cert host port in
+        (* The sock_of_stunnel has been passed to stunnel process. Close it in XAPI *)
+        ignore (Option.map (fun s -> Unixfd.safe_close s) sock_of_stunnel) ;
         (* Catch the occasional initialisation failure of stunnel: *)
         try
           let len = String.length config in
           let n =
             Unix.write Unixfd.(!config_in) (Bytes.of_string config) 0 len
           in
-          if n < len then raise Stunnel_initialisation_failed ;
-          Unixfd.safe_close config_in
+          if n < len then (
+            disconnect_with_pid ~wait:false ~force:true pid ;
+            raise Stunnel_initialisation_failed
+          ) ;
+          Unixfd.safe_close config_in ;
+          pid
         with Unix.Unix_error (err, fn, arg) ->
           write_to_log
             (Printf.sprintf
                "Caught Unix.Unix_error(%s, %s, %s); raising \
                 Stunnel_initialisation_failed"
                (Unix.error_message err) fn arg
             ) ;
+          disconnect_with_pid ~wait:false ~force:true pid ;
           raise Stunnel_initialisation_failed
     )
   in
+  let result =
+    match data_channel with
+    | `Local_host_port (h, p) ->
+        (* The stunnel will listen on a local host and port *)
+        let config = config_file ~accept:(Some (h, p)) verify_cert host port in
+        start None config
+    | `Unix_socket s ->
+        (* The stunnel will listen on a UNIX socket *)
+        let config = config_file verify_cert host port in
+        start (Some s) config
+  in
   (* Tidy up any remaining unclosed fds *)
   match result with
-  | Forkhelpers.Success (log, _) ->
-      if extended_diagnosis then (
-        write_to_log "stunnel start" ;
-        t.logfile <- log
-      ) ;
-      f t
+  | Forkhelpers.Success (log, pid) ->
+      if extended_diagnosis then write_to_log "stunnel start" ;
+      (pid, log)
   | Forkhelpers.Failure (log, exn) ->
       write_to_log ("stunnel abort: Log from stunnel: [" ^ log ^ "]") ;
-      disconnect t ;
```

Comment:
I think having the `disconnect t` here was important to avoid file descriptor leaks, otherwise the error paths in `start` need to handle disconnection which is more error prone because as we change the code and add more error paths we might forget to free it.
Was there a reason the `t` here had to be moved?

Reply:
It is moved into the `let start sock_of_stunnel config` function as `disconnect_with_pid`.
And the unix socket will be closed by `Unixfd.with_socketpair`.
The reason is just that the `t` is not applicable to the client proxy case.